### PR TITLE
Midi spanner - give spanners the possibility of rendering midi CCs [WIP]

### PR DESCRIPTION
**Pull request by hpfmn**
This is a work in progress. Currently just the switch functionality is implemented. I like to discuss if the way I'm doing is good. Also I'm not 100% happy with the typedef names.

### CODE REVIEW

**Comment by hpfmn** [comment]
Now also continuous spanners are implemented

File: libmscore/rendermidi.cpp [hunk 1]
```diff
@@ -670,6 +670,63 @@ void Score::gatherSwitchSpanners(midiCCMap &CCEvents, Spanner *s, const RepeatSe
...
+      if (startTick == -1) {
+            startTick = tick1;
+            spannerOffset = tick1 - startTick;
+            }
+
+      // Spanner ends outside this RepeatSegment
+      if (endTick == -1)
+            endTick = tick2;
+
+      int ticksPerMidiInc = s->ticks() / (s->ccEnd() - s->ccStart());
```
Comment (hpfmn) [inline libmscore/rendermidi.cpp]: not sure if that isn't accurate enough? We could also do floating point calculations and use ints at push_back?

File: libmscore/rendermidi.cpp [hunk 2]
```diff
@@ -628,61 +628,150 @@ void Score::renderStaff(EventMap* events, Staff* staff)
...
+      }
+
+//---------------------------------------------------------
 //   renderSpanners
 //---------------------------------------------------------
 
 void Score::renderSpanners(EventMap* events, int staffIdx)
       {
+      midiCCMap CCEvents = midiCCMap();
       foreach (const RepeatSegment* rs, *repeatList()) {
```
Comment (Jojo-Schmitz) [inline libmscore/rendermidi.cpp]: while touching the code, this is a good opporunity to change to C++11 `for (... : ...)`

**Comment by Jojo-Schmitz** [comment]
Seems a rebase is needed?

**Comment by hpfmn** [comment]
done ;)

**Comment by Jojo-Schmitz** [comment]
Yet another rebase needed

**Comment by Jojo-Schmitz** [comment]
yet another rebase needed

**Comment by RobFog** [comment]
Is this PR still WIP or does it just need a rebase to be merged?

**Comment by hpfmn** [comment]
I'll try to work on a rebase later today!

**Comment by hpfmn** [comment]
I guess there it could also use some work with the safe to score and save to default buttons - since they don't interact with soundbanks

**Comment by Jojo-Schmitz** [comment]
Seems another rebase is needed

**Comment by bryhoyt** [comment]
I'm very keen to know the status of this one. I'm really looking forward to it being included in the main codebase. Is there anything holding it up that could use some additional help?

**Comment by Jojo-Schmitz** [comment]
Currently it is not in a state where it could get merged, due to conflicts in the code base

**Comment by lasconic** [comment]
I plan to look at this later this year once MuseScore 2.1 will be released

**Comment by bryhoyt** [comment]
Thanks, sounds awesome!

**Comment by bryhoyt** [comment]
Not wanting to pressure you or anything, but just wondering how close you might be to looking at this one?

**Comment by Inothernews1** [comment]
Hello! As it has been a few months, what's the status on merging this with the main branch?

**Comment by Jojo-Schmitz** [comment]
Came up again in https://musescore.org/en/node/269694

Something to consider for 2.2?

**Comment by lasconic** [comment]
No, no chance in 2.2. Hopefully something similar in 3.0.

**Comment by bryhoyt** [comment]
That's a pity. I've been really hoping for it for months. Thanks for keeping it open, though.

**Comment by lasconic** [comment]
@bryhoyt just to be clear, this would completely break compatibility between release and we try to avoid this between bug fix releases. That's why this can't be put in 2.2.

**Comment by bryhoyt** [comment]
Thanks, can understand that.

**Comment by Inothernews1** [comment]
Really hoping that this will be in 3.0, it's a powerful feature we've all been hoping for!! 👍👍👍

**Comment by ericfont** [comment]
Sorry to bump, but can I assume this PR is closed?

In addition to allowing Spanners to control midi CC, I'd also like for sfz articulation to send midi CC to shape the attack.

**Comment by lasconic** [comment]
We are in the middle of the release of 2.2 so I don't have much time. 
PR is not closed. The approach here goes in the right direction. Sfz is not a spanner so it has nothing to do with this PR though. Sending MIDI CC is already possible on a staff text element (but not a dynamic...) if configured in instruments.xml. See the recent post about Hauptwerk organ here : https://musescore.org/en/node/270711

**Comment by ericfont** [comment]
Thanks for clarifying.  Sorry for getting you distracted, but the subject was just on my mind.  I should inspect this PR later and see if I can learn more.

Sfz is still related to this PR.  For one thing, sfz is most often at the start of a cresc, so I think care must be taken to make sure the implementations for sfz and cresc don't send conflicting CC messages.  The page you linked to seems to only have one CC for a staff text, but a sfz would probably have a CC for loud expression at start of note and then momentarily followed by a CC for quiet.

**Comment by lgmventura** [comment]
How is the current state of this? This should be between the highest priorities in the development of MuseScore, since the possibility of rendering midi CC opens new doors for future development and crescendo/diminuendo within single notes is a very basic and fundamental feature on playback, probably the most requested feature in the last years. Also single note pitch changes (continuous glissando) could be considered right afterwards.

**Comment by bryhoyt** [comment]
And in combination with the new support for filters in Zerberus, I could imagine some absolutely amazing things you could potentially do with this functionality.

**Comment by alago197** [comment]
> Any chance to get this into 3.0? Just got requested once again in https://musescore.org/en/node/277958
> Unfortunatly lots of merge conflicts, so a rebase won't be simple

Hi,

I would like to suggest to start with a rebase but from the previous PR, https://github.com/musescore/MuseScore/pull/2655 , which was discarded for what seemed a more general approach at the time. As this PR 2722 seems to be much more complex, perhaps starting with that previous PR would be far easier to implement in the current code? I managed to build an old 3.0 dev (from the hpfmn's "CC11"  branch) with that 2655 PR, and it worked fairly well.

**Comment by jthistle** [comment]
it might be too much work, but I'd like to take a look at rebasing this possibly in time for 3 release. Or failing that, 3.1.

**Comment by jthistle** [comment]
Or maybe not, after some advice from Marc. Perhaps it would be better to start from scratch using this and #2655 as 'inspiration'?

**Comment by jthistle** [comment]
Yes, I'd much rather rebase 200 lines of code, quickly, that I can understand well than rebase 2000 lines that I struggle to understand over a long period of time.

**Comment by alago197** [comment]
Yeah, easier.

Almost the only one problem I've encountered in hpfmn's 2655 PR is the following:  if in a measure in a score having multiple staves/instruments, there are different note lengths, the playback of the new hairping markings over single note, when applied to these staves, get messed up. I know it's a serious problem with this PR but , alas, I don't know where to look in the PR code (maybe the problem is more profound) to correct or investigate. 

I'm attaching here 2 scores written with the old 3.0 patched with PR 2655 (if you are interested, I've put this Windows build as a zip in https://drive.google.com/file/d/1xwn9fKo_sh4fBcb1lm_mGJxSNrodJbTm/view) that you can unzip in any folder (it will not mess your existing MuseScore installation).

You should open the below attached scores with this build, of course. Add in the Synthetiser the FluidR3_Mono or Musecore_General SF2 soundfonts (in fact PR 2655 worked just fine with any SF2 soundfonts I tried until now).

First score, "PR2655_test1_OK.mscx", plays well.  It's a piece I wrote last year for testing the 2655 PR. You can hear the playback as played inside this MuseScore in:  https://clyp.it/tamsnoh4?token=ebccfd8812aea05708729ae7de8fd0c1 (test ok)

But in the second score, "PR2655_test1_NOT_OK_bar1.mscx", I altered the first measure, in the violas stave, changing the first half note for a dotted quarter.  Doing so, the first stave's (violins) single-note hairpin plays well until the "ff" marking - but after it, it fails to play correctly the diminuendo hairpin that goes to the next measure:  playback at https://clyp.it/claehgto?token=e3566cba4dc138f3e6b75ae6250e3c4b
(test not ok - check bars 1-2)

In any case, , I hope you find useful tis info, @jthistle .

[PR2655_tests.zip](https://github.com/musescore/MuseScore/files/2664153/PR2655_tests.zip)

**Comment by jthistle** [comment]
Thanks! For now I'm trying to rebase it; it's taken me about an hour to do 2/3 of it. I'll be surprised if it works without a lot of refactoring :/
Also, I can't guarantee anything - I'll try my best, but if it's taking too long or becomes way out of my depth, I'm afraid I might have to leave it.

**Comment by jthistle** [comment]
Rebase completed. Doesn't work, but that's not too suprising.

**Comment by alago197** [comment]
Hi James,

Where is your rebase?? I don't see your comment below in Github. 

Alvaro
--------------------------------------------
On Tue, 12/11/18, James Thistlewood <notifications@github.com> wrote:

 Subject: Re: [musescore/MuseScore] Midi spanner - give spanners the possibility of rendering midi CCs [WIP] (#2722)
 To: "musescore/MuseScore" <MuseScore@noreply.github.com>
 "Comment" <comment@noreply.github.com>
 Date: Tuesday, December 11, 2018, 12:09 PM
 
 Rebase completed. Doesn't
 work, but that's not too suprising.
 
 —
 You are receiving this because you commented.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

**Comment by jthistle** [comment]
Sorry, I deleted that because it turned out to not be finished. It is now, and it's kind of working.

https://github.com/jthistle/MuseScore/tree/rebase-hairpin-crescendo

**Comment by lgmventura** [comment]
Wow, I tested here using the compilation for Windows through Wine on Linux and it worked very well! Thank you a lot!!! When is this going to the Master Branch? Are there still issues with it? Can someone explain the difference between both pull requests – the simpler approach and the more complex one –? I could try to give my tips according to my programming experiences.

**Comment by lgmventura** [comment]
@alago197 How do I get the hairpin to work on single notes for other instruments I add and for other files? The option is only being shown in the example files you gave.

**Comment by jthistle** [comment]
@lgmventura It's just a proof of concept thing. I'm working on a proper implementation with someone else at the moment who knows how MIDI and soundfonts work a lot better than I do. So, no, that exact code won't be going to master but we do hope to get a full version working that does the same thing :)

**Comment by alago197** [comment]
> @alago197 How do I get the hairpin to work on single notes for other instruments I add and for other files? The option is only being shown in the example files you gave on both instruments there.

Hi,

If you are talking about using the hpfmn's 2655 MuseScore build I pointed above,  you have to:

1.- Open the MusesScore program and first of all, activate the MIDI controls in Preferences (attached a screen capture)
2.- Create your score, with your instruments/staves.  Open the Mixer and dial to the left the new velocity to volume wheels (attached a screen capture)
3.- Write your music and dynamics (p, mf, etc..) and, between them, put the hairpins as desired. If you have the Inspector opened at the time you put an hairpin,  you will see there's a new checkbox , checked as default (screen attached) 
1.- Save your score as a .mscx 
2.- Edit (with WordPad for example) the .mscx, and locate the <Instrument> section you are interested. 
 In the .mscx samples I've describe above, I've used the Strings instrument.
3.- Insert the 

       <dynamics>expression</dynamics>
       <fixedVelocity>127</fixedVelocity>

  tags,  below the <InstumentId> tags

4.- Save your .mscx file with WordPad, and re-open it within MuseScore.

5.- Play it. You should see the single-note hairpins being played.

midi_controls
vel2vol_mixer_controls
hairpin_inspector

  ---
......
     
    <Part>
      <Staff id="1">
        <StaffType group="pitched">
          <name>stdNormal</name>
          </StaffType>
        <bracket type="-1" span="0"/>
        </Staff>
      <trackName>Violines</trackName>
      <Instrument>
        <longName>Violines</longName>
        <shortName>Vlns.</shortName>
        <trackName>Violines</trackName>
        <minPitchP>55</minPitchP>
        <maxPitchP>103</maxPitchP>
        <minPitchA>55</minPitchA>
        <maxPitchA>88</maxPitchA>
        <instrumentId>strings.group</instrumentId>
        <dynamics>expression</dynamics>                              <===== hpfmn's new tags here
        <fixedVelocity>127</fixedVelocity>                               <=====     "  "   "
        <Articulation>
          <velocity>100</velocity>
          <gateTime>100</gateTime>
          </Articulation>
        <Articulation name="staccatissimo">
          <velocity>100</velocity>
          <gateTime>33</gateTime>
          </Articulation>

**Comment by lgmventura** [comment]
Hi James, thanks for your answer. So, there are still issues to be resolved and points to be considered?! It would be really great to have another checkbox like "play hairpin on single note: change LPF cutoff frequency" for using in brass instruments, for example. It sounds a lot better if changing a low-pass filter cutoff frequency for certain instruments, but not for strings and some other instrument families. So it would be great to have both options. But I don't know well how MIDI and soundfonts could deal with it and how much work(around) would be necessary. Do you have a documentation on what you are actually doing?

**Comment by lgmventura** [comment]
@alago197 Thanks a lot for your detailed explanation! I'll try it soon! :-)

**Comment by alago197** [comment]
> @alago197 Thanks a lot for your detailed explanation! I'll try it soon! :-)

My pleasure. Another work I've done exclusively with hpfmn's work, almost "abusing" diminuendo/crescendo hairpins : https://clyp.it/ttxcnsiw?token=8ad2d1879babaa5b791349a38d3ecf4a .

**Comment by jthistle** [comment]
@lgmventura I'm working on an entirely new way of implementing this along with S. Christian Collins, who's made many soundfonts and knows the ins and outs of midi and soundfonts. This new implementation will be 'inspired' by this pr, in a way, but will allow other ways of creating this affect, depending on which soundfont you're using :)

**Comment by lgmventura** [comment]
Thank Alvaro @alago197 and James @jthistle ,
I've tested it here, it works pretty well when I give all the instruments the same variation in dynamics at the same time. However, if different instruments get different variations at the same time or even if there is another instrument playing anything while dynamics are being changed, it doesn't work, it just jumps from a dynamic to another. In the mixer, they're all assigned to different channels. I assume there must be some undesired coupling in the code. Or perhaps I have to give some different tag to each of them, like `<dynamics>expression_1</dynamics>` for the first instrument, `<dynamics>expression_2</dynamics>` for the second… or something like that?!
Thanks a lot, I'm really looking forward to seeing new results, especially this entirely new approach! ;-)

**Comment by alago197** [comment]
> Thank Alvaro @alago197 and James @jthistle ,
> I've tested it here, it works pretty well when I give all the instruments the same variation in dynamics at the same time. However, if different instruments get different variations at the same time or even if there is another instrument playing anything while dynamics are being changed, it doesn't work, it just jumps from a dynamic to another. In the mixer, they're all assigned to different channels. I assume there must be some undesired coupling in the code. Or perhaps I have to give some different tag (like an ID) to each of them, like `<dynamics>expression_1</dynamics>` for the first instrument, `<dynamics>expression_2</dynamics>` for the second… or something like that?!
> Thanks a lot, I'm really looking forward to seeing new results, especially this entirely new approach! ;-)

Hi,

Again, if you refer to the hpfmn's build - yes, the problem you mention seems exactly the problem I described in https://github.com/musescore/MuseScore/pull/2722#issuecomment-445904282 .  No, it's not related to having another "expression" tag, because that tag is fixed  (see https://github.com/musescore/MuseScore/pull/2655/commits/c0ca7ddc46587354a28b9bf7f26006a93f4bb20a#diff-7f889d8c5317780b684aa79a21a29625R219) .  Frankly, I'm not capable to discover where the problem really lies in the code, though.

**Comment by jthistle** [comment]
It's been a while since I've commented here. To let people know, I've added the single-note dynamic functionality in my PR #4541. I am constantly rebasing it with master to make sure it doesn't fall into the state that this PR has. 

The target for merging my PR is before MuseScore 3.0.3 is released, failing that 3.0.4. All that is required is a soundfont update from S. Christian and any needed bugfixes, plus mtest fixes which I'm leaving until last.

An detailed explanation is attached to the PR :)

**Comment by jthistle** [comment]
Single note dynamics are merged. Doesn't look like much is happening here so I'd say this could probably be closed @anatoly-os

**Pull request closed without merging**